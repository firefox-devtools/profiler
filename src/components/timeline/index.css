/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.timelineOverflowEdgeIndicatorScrollbox {
  position: absolute;

  /* if the platform scrollbar ends up being larger than --vertical-scrollbar-reserved-width (which is unexpected), make sure there is no horizontal scrollbar */
  overflow: hidden auto;
  width: 100%;
  height: 100%;
  padding-left: var(--thread-label-column-width);
  margin-right: calc(var(--vertical-scrollbar-reserved-width) * -1);
  margin-left: calc(var(--thread-label-column-width) * -1);
  background: linear-gradient(
    to left,
    white,
    #eee calc(var(--vertical-scrollbar-reserved-width) - 1px),
    var(--grey-30) 0,
    var(--grey-30) var(--vertical-scrollbar-reserved-width),
    transparent 0
  );
}

.timelineThreadList {
  width: calc(100vw - var(--vertical-scrollbar-reserved-width));
  padding: 0;
  margin: 0 0 0 calc(var(--thread-label-column-width) * -1);
  background-color: var(--grey-20);
  box-shadow: inset 0 1px var(--grey-30);
  list-style: none;
}

.timelineSettingsHiddenTracks {
  /* Make sure and override all rules that have to do with button defaults. */
  position: relative;
  width: var(--thread-label-column-width);

  /* 8px is intentional to align this text with the text above. */
  padding: 0 20px 0 8px;
  border: 0;

  /* Separator line between tracks menu button and timeline ruler */
  border-right: 1px solid var(--grey-30);
  background: none;
  color: inherit;
  font: inherit;
  text-align: left;
}

.timelineSettingsHiddenTracks:hover {
  background-color: rgb(0 0 0 / 0.1);
}

.timelineSettingsHiddenTracks:hover:active {
  background-color: rgb(0 0 0 / 0.2);
}

/* This is the dropdown arrow on the left of the button. */
.timelineSettingsHiddenTracks::before {
  position: absolute;
  top: 2px;
  right: 2px;
  border-top: 6px solid;
  border-right: 4px solid transparent;
  border-bottom: 0 solid transparent;
  border-left: 4px solid transparent;
  margin-top: 5px;
  margin-right: 5px;
  margin-left: 5px;
  color: var(--grey-60);
  content: '';

  /* Opt-out of forced colors so arrow is visible in High Contrast Mode */
  forced-color-adjust: none;
}

.timelineSettingsHiddenTracksNumber {
  font-weight: bold;
}

.timelineHeader {
  /* This should be on top of the overflow edge indicator that wraps the thread
   list, so border bottom will always be visible. */
  z-index: 1;
  display: flex;
  height: 20px;
  flex-direction: row;

  /* Timeline header content should end before the scrollbar but also border
   * bottom should extend to the whole timeline header. Because of that, we are
   * creating a padding and a minus margin with the scrollbar width to make
   * this work. */
  padding-right: var(--vertical-scrollbar-reserved-width);
  border-bottom: 1px solid var(--grey-30);
  margin-right: calc(var(--vertical-scrollbar-reserved-width) * -1);
  margin-left: calc(var(--thread-label-column-width) * -1);

  /* This div needs to be apaque so that it can properly hide the hidden tracks
   * container border below */
  background: white;
}

.tracksContainer {
  /* This should be below the timeline header. */
  z-index: 0;

  /* Moving the timeline 1 pixel below the timeline header. This way, the first
   * visible track's top border will be hidden as expected. */
  margin-top: -1px;
  margin-bottom: -1px;
}

@media (forced-colors: active) {
  .timelineSettingsHiddenTracks {
    border: 1px solid currentcolor;
    background-color: ButtonFace;
    color: ButtonText;
  }

  .timelineSettingsHiddenTracks:hover {
    background-color: SelectedItemText;
    color: SelectedItem;
  }

  .timelineSettingsHiddenTracks:hover:active {
    border-color: ButtonText;
  }

  .timelineSettingsHiddenTracks::before {
    color: currentcolor;
  }
}
