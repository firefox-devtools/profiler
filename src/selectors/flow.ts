/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';

import {
  computeProfileFlowInfo,
  computeFlowTiming,
} from '../profile-logic/marker-data';
import { getThreadSelectors } from './per-thread';
import { getActiveFlows } from './url-state';
import type { ThreadSelectors } from './per-thread';
import {
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
} from './profile';

import type {
  Selector,
  State,
  MarkerIndex,
  Marker,
  ProfileFlowInfo,
  FlowTiming,
} from 'firefox-profiler/types';

function _arraysShallowEqual(arr1: any[], arr2: any[]): boolean {
  return arr1.length === arr2.length && arr1.every((val, i) => val === arr2[i]);
}

function _createSelectorForAllThreads<T>(
  f: (selectors: ThreadSelectors, state: State) => T
): Selector<T[]> {
  let previousOutputPerThread: T[] = [];
  return function recomputeSelectorForAllThreads(state: State): T[] {
    const threads = getThreads(state);
    let outputPerThread = threads.map((_thread, i) => {
      const threadSelectors = getThreadSelectors(i);
      return f(threadSelectors, state);
    });
    if (_arraysShallowEqual(outputPerThread, previousOutputPerThread)) {
      outputPerThread = previousOutputPerThread;
    }
    previousOutputPerThread = outputPerThread;
    return outputPerThread;
  };
}

export const getFullMarkerListPerThread: Selector<Marker[][]> =
  _createSelectorForAllThreads(({ getFullMarkerList }, state) =>
    getFullMarkerList(state)
  );

export const getMarkerChartLabelGetterPerThread: Selector<
  Array<(marker: MarkerIndex) => string>
> = _createSelectorForAllThreads(({ getMarkerChartLabelGetter }, state) =>
  getMarkerChartLabelGetter(state)
);

export const getProfileFlowInfo: Selector<ProfileFlowInfo> = createSelector(
  getFullMarkerListPerThread,
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
  computeProfileFlowInfo
);

export const getFlowTiming: Selector<FlowTiming> = createSelector(
  getProfileFlowInfo,
  getActiveFlows,
  computeFlowTiming
);
