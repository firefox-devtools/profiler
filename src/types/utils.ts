/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/**
 * This type serves as documentation for how an array is meant to be used, but does
 * not support type checking. We often use an Array instead of a Map to translate
 * one type of index into another type of index. This is similar to how we use the
 * Map<K,V> type, but with the Array.
 */
export type IndexedArray<_IndexType, Value> = Array<Value>;

export type MixedObject = { [key: string]: unknown };
