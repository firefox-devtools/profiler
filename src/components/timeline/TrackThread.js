/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import classNames from 'classnames';
import explicitConnect from 'firefox-profiler/utils/connect';
import {
  withSize,
  type SizeProps,
} from 'firefox-profiler/components/shared/WithSize';
import { ThreadStackGraph } from 'firefox-profiler/components/shared/thread/StackGraph';
import { ThreadCPUGraph } from 'firefox-profiler/components/shared/thread/CPUGraph';
import { ThreadSampleGraph } from 'firefox-profiler/components/shared/thread/SampleGraph';
import { ThreadActivityGraph } from 'firefox-profiler/components/shared/thread/ActivityGraph';

import {
  getProfileInterval,
  getCommittedRange,
  getCategories,
  getSelectedThreadIndexes,
  getTimelineType,
  getThreadSelectorsFromThreadsKey,
  getIsExperimentalCPUGraphsEnabled,
  getImplementationFilter,
  getZeroAt,
} from 'firefox-profiler/selectors';
import {
  TimelineMarkersJank,
  TimelineMarkersFileIo,
  TimelineMarkersOverview,
  TimelineMarkersMemory,
} from './Markers';
import {
  updatePreviewSelection,
  changeSelectedCallNode,
  focusCallTree,
  selectSelfCallNode,
} from 'firefox-profiler/actions/profile-view';
import { reportTrackThreadHeight } from 'firefox-profiler/actions/app';
import { EmptyThreadIndicator } from './EmptyThreadIndicator';
import { getTrackSelectionModifiers } from 'firefox-profiler/utils';
import './TrackThread.css';

import type {
  TimelineType,
  Thread,
  ThreadIndex,
  CategoryList,
  IndexIntoSamplesTable,
  Milliseconds,
  StartEndRange,
  ImplementationFilter,
  IndexIntoCallNodeTable,
  SelectedState,
  State,
  ThreadsKey,
} from 'firefox-profiler/types';
import type { CallNodeInfo } from 'firefox-profiler/profile-logic/call-node-info';

import type { ConnectedProps } from 'firefox-profiler/utils/connect';

type OwnProps = {|
  +threadsKey: ThreadsKey,
  +trackType: 'expanded' | 'condensed',
  +showMemoryMarkers?: boolean,
  +trackName: string,
|};

type StateProps = {|
  +fullThread: Thread,
  +rangeFilteredThread: Thread,
  +filteredThread: Thread,
  +callNodeInfo: CallNodeInfo,
  +unfilteredSamplesRange: StartEndRange | null,
  +interval: Milliseconds,
  +rangeStart: Milliseconds,
  +rangeEnd: Milliseconds,
  +sampleIndexOffset: number,
  +categories: CategoryList,
  +timelineType: TimelineType,
  +hasFileIoMarkers: boolean,
  +samplesSelectedStates: null | SelectedState[],
  +sampleNonInvertedCallNodes: Array<IndexIntoCallNodeTable | null>,
  +treeOrderSampleComparator: (
    IndexIntoSamplesTable,
    IndexIntoSamplesTable
  ) => number,
  +selectedThreadIndexes: Set<ThreadIndex>,
  +enableCPUUsage: boolean,
  +isExperimentalCPUGraphsEnabled: boolean,
  +implementationFilter: ImplementationFilter,
  +callTreeVisible: boolean,
  +zeroAt: Milliseconds,
|};

type DispatchProps = {|
  +updatePreviewSelection: typeof updatePreviewSelection,
  +changeSelectedCallNode: typeof changeSelectedCallNode,
  +focusCallTree: typeof focusCallTree,
  +selectSelfCallNode: typeof selectSelfCallNode,
  +reportTrackThreadHeight: typeof reportTrackThreadHeight,
|};

type Props = {|
  ...SizeProps,
  ...ConnectedProps<OwnProps, StateProps, DispatchProps>,
|};

class TimelineTrackThreadImpl extends PureComponent<Props> {
  /**
   * Handle when a sample is clicked in the ThreadStackGraph and in the ThreadActivityGraph.
   * This will select the leaf-most stack frame or call node.
   */
  _onSampleClick = (
    event: SyntheticMouseEvent<>,
    sampleIndex: IndexIntoSamplesTable | null
  ) => {
    const modifiers = getTrackSelectionModifiers(event);
    if (modifiers.ctrlOrMeta || modifiers.shift) {
      // Do nothing, the track selection logic will kick in.
      return;
    }

    const {
      threadsKey,
      selectSelfCallNode,
      focusCallTree,
      selectedThreadIndexes,
      callTreeVisible,
    } = this.props;

    // Sample clicking only works for one thread. See issue #2709
    if (selectedThreadIndexes.size === 1) {
      selectSelfCallNode(threadsKey, sampleIndex);

      if (sampleIndex !== null && callTreeVisible) {
        // If the user clicked outside of the activity graph (sampleIndex === null),
        // then we don't need to focus the call tree. This action also selects
        // the call tree panel, which we don't want either in this case.
        focusCallTree();
      }
    }
    if (
      typeof threadsKey === 'number' &&
      selectedThreadIndexes.has(threadsKey)
    ) {
      // We could have multiple threads selected here, and we wouldn't want
      // to de-select one when interacting with it.
      event.stopPropagation();
    }
  };

  _onMarkerSelect = (start: Milliseconds, end: Milliseconds) => {
    const { rangeStart, rangeEnd, updatePreviewSelection } = this.props;
    updatePreviewSelection({
      hasSelection: true,
      isModifying: false,
      selectionStart: Math.max(rangeStart, start),
      selectionEnd: Math.min(rangeEnd, end),
    });
  };

  componentDidUpdate() {
    const { threadsKey, height, reportTrackThreadHeight } = this.props;
    // Most likely this track height shouldn't change, but if it does, report it.
    // The action will only dispatch on changed values.
    reportTrackThreadHeight(threadsKey, height);
  }

  render() {
    const {
      fullThread,
      filteredThread,
      rangeFilteredThread,
      threadsKey,
      interval,
      rangeStart,
      rangeEnd,
      sampleIndexOffset,
      callNodeInfo,
      unfilteredSamplesRange,
      categories,
      timelineType,
      hasFileIoMarkers,
      showMemoryMarkers,
      sampleNonInvertedCallNodes,
      samplesSelectedStates,
      treeOrderSampleComparator,
      trackType,
      trackName,
      enableCPUUsage,
      isExperimentalCPUGraphsEnabled,
      implementationFilter,
      zeroAt,
    } = this.props;

    const processType = filteredThread.processType;
    const displayJank = processType !== 'plugin';
    const displayMarkers =
      (filteredThread.showMarkersInTimeline ||
        filteredThread.name === 'GeckoMain' ||
        filteredThread.name === 'Compositor' ||
        filteredThread.name === 'Renderer' ||
        filteredThread.name === 'AndroidUI (JVM)' ||
        filteredThread.name === 'CrRendererMain' ||
        filteredThread.name === 'Merged thread' ||
        filteredThread.name.startsWith('MediaDecoderStateMachine')) &&
      processType !== 'plugin';

    return (
      <div className={classNames('timelineTrackThread', trackType)}>
        {trackType === 'expanded' && showMemoryMarkers ? (
          <TimelineMarkersMemory
            rangeStart={rangeStart}
            rangeEnd={rangeEnd}
            threadsKey={threadsKey}
            onSelect={this._onMarkerSelect}
          />
        ) : null}
        {trackType === 'expanded' && hasFileIoMarkers ? (
          <TimelineMarkersFileIo
            rangeStart={rangeStart}
            rangeEnd={rangeEnd}
            threadsKey={threadsKey}
            onSelect={this._onMarkerSelect}
          />
        ) : null}
        {displayJank ? (
          <TimelineMarkersJank
            rangeStart={rangeStart}
            rangeEnd={rangeEnd}
            threadsKey={threadsKey}
            onSelect={this._onMarkerSelect}
          />
        ) : null}
        {trackType === 'expanded' && displayMarkers ? (
          <TimelineMarkersOverview
            rangeStart={rangeStart}
            rangeEnd={rangeEnd}
            threadsKey={threadsKey}
            onSelect={this._onMarkerSelect}
          />
        ) : null}

        {(timelineType === 'category' || timelineType === 'cpu-category') &&
        !filteredThread.isJsTracer ? (
          <>
            <ThreadActivityGraph
              className="threadActivityGraph"
              trackName={trackName}
              interval={interval}
              fullThread={fullThread}
              rangeFilteredThread={rangeFilteredThread}
              rangeStart={rangeStart}
              rangeEnd={rangeEnd}
              sampleIndexOffset={sampleIndexOffset}
              onSampleClick={this._onSampleClick}
              categories={categories}
              samplesSelectedStates={samplesSelectedStates}
              treeOrderSampleComparator={treeOrderSampleComparator}
              enableCPUUsage={enableCPUUsage}
              implementationFilter={implementationFilter}
              timelineType={timelineType}
              zeroAt={zeroAt}
            />
            {trackType === 'expanded' ? (
              <ThreadSampleGraph
                className="threadSampleGraph"
                trackName={trackName}
                interval={interval}
                thread={filteredThread}
                rangeStart={rangeStart}
                rangeEnd={rangeEnd}
                samplesSelectedStates={samplesSelectedStates}
                categories={categories}
                onSampleClick={this._onSampleClick}
                timelineType={timelineType}
                implementationFilter={implementationFilter}
                zeroAt={zeroAt}
              />
            ) : null}
            {isExperimentalCPUGraphsEnabled &&
            rangeFilteredThread.samples.threadCPURatio !== undefined ? (
              <ThreadCPUGraph
                className="threadCPUGraph"
                trackName={trackName}
                interval={interval}
                thread={filteredThread}
                rangeStart={rangeStart}
                rangeEnd={rangeEnd}
                callNodeInfo={callNodeInfo}
                samplesSelectedStates={samplesSelectedStates}
                categories={categories}
                onSampleClick={this._onSampleClick}
              />
            ) : null}
          </>
        ) : (
          <ThreadStackGraph
            className="threadStackGraph"
            trackName={trackName}
            interval={interval}
            thread={filteredThread}
            rangeStart={rangeStart}
            rangeEnd={rangeEnd}
            callNodeInfo={callNodeInfo}
            sampleNonInvertedCallNodes={sampleNonInvertedCallNodes}
            samplesSelectedStates={samplesSelectedStates}
            categories={categories}
            onSampleClick={this._onSampleClick}
          />
        )}
        <EmptyThreadIndicator
          thread={filteredThread}
          interval={interval}
          rangeStart={rangeStart}
          rangeEnd={rangeEnd}
          unfilteredSamplesRange={unfilteredSamplesRange}
        />
      </div>
    );
  }
}

export const TimelineTrackThread = explicitConnect<
  OwnProps,
  StateProps,
  DispatchProps,
>({
  mapStateToProps: (state: State, ownProps: OwnProps) => {
    const { threadsKey } = ownProps;
    const selectors = getThreadSelectorsFromThreadsKey(threadsKey);
    const selectedThreadIndexes = getSelectedThreadIndexes(state);
    const committedRange = getCommittedRange(state);
    const fullThread = selectors.getThread(state);
    const timelineType = getTimelineType(state);
    const enableCPUUsage =
      timelineType === 'cpu-category' &&
      fullThread.samples.threadCPURatio !== undefined;

    return {
      fullThread,
      filteredThread: selectors.getFilteredThread(state),
      rangeFilteredThread: selectors.getRangeFilteredThread(state),
      callNodeInfo: selectors.getCallNodeInfo(state),
      sampleNonInvertedCallNodes:
        selectors.getSampleIndexToNonInvertedCallNodeIndexForFilteredThread(
          state
        ),
      unfilteredSamplesRange: selectors.unfilteredSamplesRange(state),
      interval: getProfileInterval(state),
      rangeStart: committedRange.start,
      rangeEnd: committedRange.end,
      sampleIndexOffset: selectors.getFilteredSampleIndexOffset(state),
      categories: getCategories(state),
      timelineType,
      hasFileIoMarkers:
        selectors.getTimelineFileIoMarkerIndexes(state).length !== 0,
      samplesSelectedStates:
        selectors.getSamplesSelectedStatesInFilteredThread(state),
      treeOrderSampleComparator:
        selectors.getTreeOrderComparatorInFilteredThread(state),
      selectedThreadIndexes,
      enableCPUUsage,
      isExperimentalCPUGraphsEnabled: getIsExperimentalCPUGraphsEnabled(state),
      implementationFilter: getImplementationFilter(state),
      callTreeVisible: selectors.getUsefulTabs(state).includes('calltree'),
      zeroAt: getZeroAt(state),
    };
  },
  mapDispatchToProps: {
    updatePreviewSelection,
    changeSelectedCallNode,
    focusCallTree,
    selectSelfCallNode,
    reportTrackThreadHeight,
  },
  component: withSize<Props>(TimelineTrackThreadImpl),
});
