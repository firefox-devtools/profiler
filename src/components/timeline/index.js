/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import { FullTimeline } from 'firefox-profiler/components/timeline/FullTimeline';

type TimelineProps = {||};

export class Timeline extends PureComponent<TimelineProps> {
  // This may contain a function that's called whenever we want to remove the
  // "wheel" listener.
  _removeWheelListener: null | (() => mixed) = null;

  // This effectively disable the pinch-to-zoom as well as ctrl+mousewheel
  // gestures. Indeed in the timeline it is confusing. In the future we'll want
  // to couple this with the preview selection like in the Viewport HOC.
  preventPinchToZoom(e: WheelEvent) {
    if (e.ctrlKey) {
      e.preventDefault();
    }
  }

  // This will be registered as a ref property to the DOM element displaying the
  // tracks. We use this solution of registering the wheel event in this ref
  // listener because:
  // * we can't use React's event handling, because it doesn't allow us to use
  //   the "passive: false" way of registering the event handler. But we need this
  //   if we want to be able to prevent the default action of page zooming.
  // * we want to be sure to register it whenever the element changes.
  _onTimelineMountWithRef = (ref: HTMLElement | null) => {
    if (this._removeWheelListener) {
      this._removeWheelListener();
      this._removeWheelListener = null;
    }

    if (!ref) {
      return;
    }

    // without pinning to a const variable, Flow isn't sure that we don't change
    // the `ref` variable in some of the function calls below, and therefore
    // that it won't be null.
    const existingRef = ref;

    // Disable pinch-to-zoom and ctrl + wheel otherwise, on the timeline.
    // Indeed the users are used to this gesture  to zoom in in our charts, and
    // may use the same gesture elsewhere because of their habits, however this
    // doesn't do what they expect and instead zooms in the page, which is distracting.
    existingRef.addEventListener('wheel', this.preventPinchToZoom, {
      passive: false,
    });

    this._removeWheelListener = () => {
      existingRef.removeEventListener('wheel', this.preventPinchToZoom, {
        passive: false,
      });
    };
  };

  componentWillUnmount() {
    if (this._removeWheelListener) {
      this._removeWheelListener();
      this._removeWheelListener = null;
    }
  }

  render() {
    return <FullTimeline innerElementRef={this._onTimelineMountWithRef} />;
  }
}
