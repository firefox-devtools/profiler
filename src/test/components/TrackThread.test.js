/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import type { Profile, FileIoPayload } from 'firefox-profiler/types';

import * as React from 'react';
import { Provider } from 'react-redux';
import { oneLine } from 'common-tags';

import { render, act } from 'firefox-profiler/test/fixtures/testing-library';
import {
  changeTimelineType,
  changeInvertCallstack,
  changeSelectedCallNode,
} from '../../actions/profile-view';
import { TimelineTrackThread } from '../../components/timeline/TrackThread';
import { getPreviewSelection } from '../../selectors/profile';
import { selectedThreadSelectors } from '../../selectors/per-thread';
import { ensureExists } from '../../utils/flow';

import {
  autoMockCanvasContext,
  flushDrawLog,
} from '../fixtures/mocks/canvas-context';
import { mockRaf } from '../fixtures/mocks/request-animation-frame';
import { storeWithProfile } from '../fixtures/stores';
import {
  addRootOverlayElement,
  removeRootOverlayElement,
  fireFullClick,
} from '../fixtures/utils';
import {
  getProfileFromTextSamples,
  getProfileWithMarkers,
} from '../fixtures/profiles/processed-profile';
import { autoMockElementSize } from '../fixtures/mocks/element-size';
import { autoMockIntersectionObserver } from '../fixtures/mocks/intersection-observer';

// The graph is 400 pixels wide based on the element size mock. Each stack is
// 100 pixels wide. Use the value 50 to click in the middle of this stack, and
// incrementing by steps of 100 pixels to get to the next stack.
const GRAPH_WIDTH = 400;
const GRAPH_HEIGHT = 50;

/**
 * This test is asserting behavior more for the ThreadStackGraph component. The
 * ThreadActivityGraph component was added as a new default. Currently this test
 * only checks the older behavior.
 */
describe('timeline/TrackThread', function () {
  beforeEach(addRootOverlayElement);
  afterEach(removeRootOverlayElement);
  autoMockCanvasContext();
  autoMockElementSize({ width: GRAPH_WIDTH, height: GRAPH_HEIGHT });
  autoMockIntersectionObserver();

  function getSamplesProfile() {
    return getProfileFromTextSamples(`
      a  d  g  j
      b  e  h  k
      c  f  i  l
    `).profile;
  }

  function getMarkersProfile(
    testMarkers = [
      ['Marker A', 0],
      ['Marker B', 1],
      ['Marker C', 2],
      ['Marker D', 3],
    ]
  ) {
    const profile = getProfileWithMarkers(testMarkers);
    const [thread] = profile.threads;
    thread.name = 'GeckoMain';
    thread.isMainThread = true;
    thread.processType = 'default';
    return profile;
  }

  function setup(profile: Profile) {
    const store = storeWithProfile(profile);
    const { getState, dispatch } = store;
    const threadIndex = 0;
    const flushRafCalls = mockRaf();

    type Coordinate = {| pageX: number, pageY: number |};

    // Look through the draw log and find the center of a specific fillRect
    // call. This is a good way to know where the canvas drew something.
    function getFillRectCenterByIndex(log: any[], index: number): Coordinate {
      type FillRectCall = [string, number, number, number, number];
      const calls: FillRectCall[] = log.filter(
        (call) => call[0] === 'fillRect'
      );
      const call = calls[index];
      if (!call) {
        console.error(log);
        throw new Error(`Could not find a fillRect call at ${index}.`);
      }
      const [, x, y, w, h] = call;
      return { pageX: x + w * 0.5, pageY: y + h * 0.5 };
    }

    // Note: These tests were first written with the timeline using the ThreadStackGraph.
    // This is not the default view, so dispatch an action to change to the older default
    // view.
    store.dispatch(changeTimelineType('stack'));

    const renderResult = render(
      <Provider store={store}>
        <TimelineTrackThread
          threadsKey={threadIndex}
          trackType="expanded"
          trackName="Test Track"
        />
      </Provider>
    );
    const { container } = renderResult;

    // WithSize uses requestAnimationFrame
    flushRafCalls();

    const stackGraphCanvas = () =>
      ensureExists(
        container.querySelector('.threadStackGraphCanvas'),
        `Couldn't find the stack graph canvas, with selector .threadStackGraphCanvas`
      );
    const markerCanvas = () =>
      ensureExists(
        container.querySelector(oneLine`
          .timelineMarkersGeckoMain
          .timelineMarkersCanvas
        `),
        `Couldn't find the marker canvas`
      );

    return {
      ...renderResult,
      dispatch,
      getState,
      profile,
      thread: profile.threads[0],
      store,
      threadIndex,
      stackGraphCanvas,
      markerCanvas,
      getFillRectCenterByIndex,
    };
  }

  it('matches the snapshot for the component', () => {
    const { container } = setup(getSamplesProfile());
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches the 2d canvas draw snapshot', () => {
    setup(getSamplesProfile());
    expect(flushDrawLog()).toMatchSnapshot();
  });

  it('can click a stack in the stack graph in normal call trees', function () {
    const {
      getState,
      stackGraphCanvas,
      profile,
      thread,
      getFillRectCenterByIndex,
    } = setup(getSamplesProfile());

    const log = flushDrawLog();

    // Provide a quick helper for nicely asserting the call node path.
    const getCallNodePath = () =>
      selectedThreadSelectors
        .getSelectedCallNodePath(getState())
        .map(
          (funcIndex) =>
            profile.shared.stringArray[thread.funcTable.name[funcIndex]]
        );

    fireFullClick(stackGraphCanvas(), getFillRectCenterByIndex(log, 0));
    expect(getCallNodePath()).toEqual(['a', 'b', 'c']);

    fireFullClick(stackGraphCanvas(), getFillRectCenterByIndex(log, 1));
    expect(getCallNodePath()).toEqual(['d', 'e', 'f']);

    fireFullClick(stackGraphCanvas(), getFillRectCenterByIndex(log, 2));
    expect(getCallNodePath()).toEqual(['g', 'h', 'i']);

    fireFullClick(stackGraphCanvas(), getFillRectCenterByIndex(log, 3));
    expect(getCallNodePath()).toEqual(['j', 'k', 'l']);
  });

  it('can click a stack in the stack graph in inverted call trees', function () {
    const {
      dispatch,
      getState,
      stackGraphCanvas,
      profile,
      thread,
      getFillRectCenterByIndex,
    } = setup(getSamplesProfile());

    // Provide a quick helper for nicely asserting the call node path.
    const getCallNodePath = () =>
      selectedThreadSelectors
        .getSelectedCallNodePath(getState())
        .map(
          (funcIndex) =>
            profile.shared.stringArray[thread.funcTable.name[funcIndex]]
        );

    function changeInvertCallstackAndGetDrawLog(value) {
      // We don't want a selected stack graph to change fillRect ordering.
      act(() => {
        dispatch(changeSelectedCallNode(0, []));
      });
      flushDrawLog();
      act(() => {
        dispatch(changeInvertCallstack(value));
      });
      return flushDrawLog();
    }

    // Switch to "inverted" mode to test with this state
    {
      const log = changeInvertCallstackAndGetDrawLog(true);

      fireFullClick(stackGraphCanvas(), getFillRectCenterByIndex(log, 0));
      expect(getCallNodePath()).toEqual(['c']);

      fireFullClick(stackGraphCanvas(), getFillRectCenterByIndex(log, 2));
      expect(getCallNodePath()).toEqual(['i']);
    }
    {
      // Switch back to "uninverted" mode
      const log = changeInvertCallstackAndGetDrawLog(false);

      fireFullClick(stackGraphCanvas(), getFillRectCenterByIndex(log, 0));
      expect(getCallNodePath()).toEqual(['a', 'b', 'c']);

      fireFullClick(stackGraphCanvas(), getFillRectCenterByIndex(log, 2));
      expect(getCallNodePath()).toEqual(['g', 'h', 'i']);
    }
  });

  it('can click a marker', function () {
    const { getState, markerCanvas, getFillRectCenterByIndex } = setup(
      getMarkersProfile([
        ['DOMEvent', 0, 4],
        ['DOMEvent', 4, 8],
      ])
    );

    const log = flushDrawLog();

    function clickAndGetMarkerName(event) {
      fireFullClick(markerCanvas(), event);
      return getPreviewSelection(getState());
    }

    // Currently markers are drawn with 3 fillRects, the middle of the three is the
    // big interesting one. If this test breaks, likely the drawing strategy
    // has changed.
    const determineIndex = (i) => i * 3 + 1;

    expect(
      clickAndGetMarkerName(getFillRectCenterByIndex(log, determineIndex(0)))
    ).toMatchObject({
      selectionStart: 0,
      selectionEnd: 4,
    });

    expect(
      clickAndGetMarkerName(getFillRectCenterByIndex(log, determineIndex(1)))
    ).toMatchObject({
      selectionStart: 4,
      selectionEnd: 8,
    });
  });

  it('does not add disk io markers if none are present', function () {
    const noMarkers = [];
    const { queryByTestId } = setup(getMarkersProfile(noMarkers));
    expect(queryByTestId('TimelineMarkersFileIo')).not.toBeInTheDocument();
  });

  it('adds file io markers if they are present', function () {
    const fileIoMarker = [
      [
        'FileIO',
        2,
        3,
        ({
          type: 'FileIO',
          source: 'PoisionOIInterposer',
          filename: '/foo/bar/',
          operation: 'read/write',
        }: FileIoPayload),
      ],
    ];
    const { getByTestId } = setup(getMarkersProfile(fileIoMarker));
    expect(getByTestId('TimelineMarkersFileIo')).toBeInTheDocument();
  });

  it('does not add off-thread file io markers even if they are present', function () {
    const fileIoMarker = [
      [
        'FileIO',
        2,
        3,
        ({
          type: 'FileIO',
          source: 'PoisionOIInterposer',
          filename: '/foo/bar/',
          operation: 'read/write',
          threadId: 123,
        }: FileIoPayload),
      ],
    ];
    const { queryByTestId } = setup(getMarkersProfile(fileIoMarker));
    expect(queryByTestId('TimelineMarkersFileIo')).not.toBeInTheDocument();
  });
});
