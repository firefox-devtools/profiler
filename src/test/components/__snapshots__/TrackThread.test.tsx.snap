// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`timeline/TrackThread matches the 2d canvas draw snapshot 1`] = `
Array [
  Array [
    "set fillStyle",
    "#003eaa",
  ],
  Array [
    "fillRect",
    0,
    0,
    100,
    50,
  ],
  Array [
    "fillRect",
    100,
    0,
    100,
    50,
  ],
  Array [
    "fillRect",
    200,
    0,
    100,
    50,
  ],
  Array [
    "fillRect",
    300,
    0,
    100,
    50,
  ],
  Array [
    "clearRect",
    0,
    0,
    400,
    50,
  ],
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "scale",
    1,
    1,
  ],
]
`;

exports[`timeline/TrackThread matches the snapshot for the component 1`] = `
<div
  class="timelineTrackThread expanded"
>
  <div
    class="timelineMarkers selected"
    data-testid="TimelineMarkersJank"
  >
    <div
      class="react-contextmenu-wrapper"
    >
      <div>
        <canvas
          class="timelineMarkersCanvas"
          height="50"
          width="400"
        />
      </div>
    </div>
  </div>
  <div
    class="threadStackGraph"
  >
    <canvas
      class="threadStackGraphCanvas threadHeightGraphCanvas"
      height="50"
      width="400"
    >
      <h2>
        Stack Graph for 
        Test Track
      </h2>
      <p>
        This graph charts the stack height of each sample.
      </p>
    </canvas>
  </div>
  <div
    class="timelineEmptyThreadIndicator"
  />
</div>
`;
